### PR TITLE
fix: testnet relay

### PR DESCRIPTION
**Pull request by jjyr**
Refactoring ugly code,
Add `is_bad_tx` function on `PoolError` and `TransactionError`, use this method to detect a tx is intended bad tx or just caused by the different tip.

### CODE REVIEW

**Comment by doitian** [comment]
Please provide a description in the PR body to describe the reason of the changes.

File: sync/src/relayer/transaction_process.rs [hunk 1]
```diff
@@ -77,13 +77,20 @@ impl<'a, CS: ChainStore> TransactionProcess<'a, CS> {
...
                 }
             }
-            Err(PoolError::InvalidTx(TransactionError::Unknown))
-            | Err(PoolError::InvalidTx(TransactionError::Conflict))
-            | Err(PoolError::InvalidTx(TransactionError::Immature))
-            | Err(PoolError::InvalidTx(TransactionError::CellbaseImmaturity)) => {
-                // this error may occured when peer's tip is different with us,
-                // we can't proof peer is bad so just ignore this
-                debug!(target: "relay", "peer {} relay a conflict or unknown input / dep tx: {:?}", self.peer, tx);
+            Err(PoolError::InvalidTx(TransactionError::NullInput))
```
Comment (doitian) [inline sync/src/relayer/transaction_process.rs]: This looks ugly. It is better to add a function on `TransactionError` or `PoolError` to tell whether the peer should be banned.

File: sync/src/relayer/transaction_process.rs [hunk 2]
```diff
@@ -77,13 +77,20 @@ impl<'a, CS: ChainStore> TransactionProcess<'a, CS> {
...
+            | Err(PoolError::InvalidTx(TransactionError::Empty))
+            | Err(PoolError::InvalidTx(TransactionError::OutputsSumOverflow))
+            | Err(PoolError::InvalidTx(TransactionError::InvalidScript))
+            | Err(PoolError::InvalidTx(TransactionError::ScriptFailure(_)))
+            | Err(PoolError::InvalidTx(TransactionError::InvalidSignature))
+            | Err(PoolError::InvalidTx(TransactionError::InvalidValidSince))
+            | Err(PoolError::Cellbase) => {
+                debug!(target: "relay", "peer {} relay a invalid tx: {:?}", self.peer, tx);
+                // TODO use report score interface
+                self.nc.ban_peer(self.peer, DEFAULT_BAN_TIME);
```
Comment (doitian) [inline sync/src/relayer/transaction_process.rs]: We can capture the exception here using sentry to ease debugging.

File: sync/src/relayer/transaction_process.rs [hunk 3]
```diff
@@ -95,9 +85,19 @@ impl<'a, CS: ChainStore> TransactionProcess<'a, CS> {
...
+                    sentry::capture_message(
+                        &format!(
+                            "ban peer {} {:?}, reason: relay invalid tx: {:?}, error: {:?}",
+                            self.peer, DEFAULT_BAN_TIME, tx, err
+                        ),
+                        sentry::Level::Info,
+                    );
+                    self.nc.ban_peer(self.peer, DEFAULT_BAN_TIME);
+                } else {
+                    debug!(target: "relay", "peer {} relay a conflict or missing input tx: {:?}, error: {:?}", self.peer, tx, err);
```
Comment (TheWaWaR) [inline sync/src/relayer/transaction_process.rs]: Should not print tx content, it may very large.

File: sync/src/relayer/transaction_process.rs [hunk 4]
```diff
@@ -75,33 +73,28 @@ impl<'a, CS: ChainStore> TransactionProcess<'a, CS> {
...
-                // this error may occured when peer's tip is different with us,
-                // we can't proof peer is bad so just ignore this
-                debug!(target: "relay", "peer {} relay tx verify err: {:?}, error: {:?}", self.peer, tx, err);
+                if err.is_bad_tx() {
+                    debug!(target: "relay", "peer {} relay a invalid tx: {:?}, error: {:?}", self.peer, tx_hash, err);
+                    sentry::capture_message(
+                        &format!(
+                            "ban peer {} {:?}, reason: relay invalid tx: {:?}, error: {:?}",
+                            self.peer, DEFAULT_BAN_TIME, tx, err
+                        ),
```
Comment (TheWaWaR) [inline sync/src/relayer/transaction_process.rs]: I wonder what will happen, when the transaction(tx) is large, and the network to sentry.io is bad?

**Pull request merged**